### PR TITLE
Implement GasRouter

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

- Adds `GasRouter` with remote router handle gas overhead mapping in storage

### Drive-by changes

- Updates `hyperlane-token` to use `GasRouter`

### Related issues

- Fixes https://github.com/hyperlane-xyz/hyperlane-token/issues/27

### Backward compatibility

No

### Testing

Unit Tests

### CODE REVIEW

**Comment by yorhodes** [comment]
> Is this ready for review yet?

yes

File: solidity/contracts/GasRouter.sol [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        for (uint256 i = 0; i < gasConfigs.length; i += 1) {
+            setGasOverheadConfig(gasConfigs[i]);
+        }
+    }
+
+    /**
+     * @notice Enroll a remote router with a handle gas overhead
+     * @param gasConfig The GasRouterConfig struct
+     */
+    function setGasOverheadConfig(GasRouterConfig calldata gasConfig)
```
Comment (nambrot) [inline solidity/contracts/GasRouter.sol:41]: I know we have been doing this, but should we even have the single versions of this and just keep the batch versions?

File: solidity/contracts/GasRouter.sol [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        public
+        onlyOwner
+    {
+        _setGasOverhead(gasConfig.domain, gasConfig.handleGasOverhead);
+    }
+
+    function setGasOverhead(uint32 domain, uint256 gas) external onlyOwner {
```
Comment (nambrot) [inline solidity/contracts/GasRouter.sol:48]: I also know that i have been advocating for non-struct versions of calls, but i wonder if this is a bit superflous and just adding to the bytecode

File: typescript/sdk/src/deploy/router/GasRouterDeployer.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      );
+      const remoteConfigs = remoteDomains.map((domain, i) => ({
+        domain,
+        handleGasOverhead: this.configMap[remoteChains[i]].handleGasOverhead,
+      }));
+      this.logger(
+        `Enroll remote (${remoteChains}) handle gas overhead on ${local}`,
+      );
+      const chainConnection = this.multiProvider.getChainConnection(local);
+      await chainConnection.handleTx(
```
Comment (nambrot) [inline typescript/sdk/src/deploy/router/GasRouterDeployer.ts:53]: Optional Nit: Check that the values have not already been set this way, otherwise skip

**Review by nambrot — APPROVED** [review]
Mostly nits on my side

File: solidity/contracts/GasRouter.sol [hunk 4]
```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.6.11;
+
+import {Router} from "./Router.sol";
+
+abstract contract GasRouter is Router {
+    // ============ Mutable Storage ============
+    mapping(uint32 => uint256) public handleGasOverhead;
```
Comment (tkporter) [inline solidity/contracts/GasRouter.sol:8]: I'm wondering if we should pick a different name here, particularly around "handle" & "overhead"

My reasoning is that it's only handle gas if the contract inheriting from `GasRouter` happens to be using the DefaultIsmInterchainGasPaymaster. If they happen to be using the base InterchainGasPaymaster, this doesn't quite apply

and then overhead feels like it's being used slightly differently from when I've been using it - I've been thinking of it as some kind of gas cost that's intrinsic and on top of some other application-specific gas. In this situation, this feels like "overhead" is used to refer to the application-specific gas

maybe something like `destinationGas` would be more general?

File: solidity/contracts/GasRouter.sol [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+     */
+    event HandleGasOverheadSet(uint32 indexed domain, uint256 gas);
+
+    struct GasRouterConfig {
+        uint32 domain;
+        uint256 handleGasOverhead;
+    }
+
+    /**
+     * @notice Batch version of `enrollRemoteRouterWithGas`
```
Comment (tkporter) [inline solidity/contracts/GasRouter.sol:25]: nit: this seems stale

File: solidity/contracts/GasRouter.sol [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        external
+        onlyOwner
+    {
+        for (uint256 i = 0; i < gasConfigs.length; i += 1) {
+            setGasOverheadConfig(gasConfigs[i]);
+        }
+    }
+
+    /**
+     * @notice Enroll a remote router with a handle gas overhead
```
Comment (tkporter) [inline solidity/contracts/GasRouter.sol:38]: nit: also seems stale

File: solidity/contracts/GasRouter.sol (hunk 2 above)
Comment (tkporter) [inline solidity/contracts/GasRouter.sol:48]: natspec for these other functions would be nice

File: solidity/contracts/GasRouter.sol [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        onlyOwner
+    {
+        _setGasOverhead(gasConfig.domain, gasConfig.handleGasOverhead);
+    }
+
+    function setGasOverhead(uint32 domain, uint256 gas) external onlyOwner {
+        _setGasOverhead(domain, gas);
+    }
+
+    function quoteGasPayment(uint32 _destinationDomain)
```
Comment (tkporter) [inline solidity/contracts/GasRouter.sol:52]: this is slick

File: solidity/contracts/GasRouter.sol [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            _dispatchWithGas(
+                _destinationDomain,
+                _messageBody,
+                handleGasOverhead[_destinationDomain],
+                _gasPayment,
+                _gasPaymentRefundAddress
+            );
+    }
+
+    function _dispatchWithGas(
```
Comment (tkporter) [inline solidity/contracts/GasRouter.sol:85]: definitely would like some natspec on this one in particular

File: typescript/sdk/src/deploy/router/GasRouterDeployer.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+          domain,
+          gas: this.configMap[remoteChains[i]].gas,
+        }))
+        .filter(({ gas }, index) => !currentConfigs[index].eq(gas));
+      this.logger(
+        `Enroll remote (${remoteChains}) handle gas overhead on ${local}`,
+      );
+      const chainConnection = this.multiProvider.getChainConnection(local);
+      await chainConnection.handleTx(
+        contracts.router.setDestinationGas(remoteConfigs),
```
Comment (nambrot) [inline typescript/sdk/src/deploy/router/GasRouterDeployer.ts:59]: I think you still want to skip here if remoteConfigs is empty?

File: typescript/sdk/src/router.ts [hunk 10]
```diff
@@ -27,3 +31,34 @@ export type ConnectionClientConfig = {
...
+      ),
+    );
+
+  getOwners = () =>
+    promiseObjAll(
+      objMap(this.contractsMap, (_, contracts) => contracts.router.owner()),
+    );
+}
+
+export class GasRouterApp<
```
Comment (nambrot) [inline typescript/sdk/src/router.ts:52]: Nice

**Review by nambrot — APPROVED** [review]
Just minor comment

File: typescript/sdk/src/deploy/router/GasRouterDeployer.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+      ...options,
+    });
+  }
+
+  async enrollRemoteRouters(
+    contractsMap: ChainMap<Chain, Contracts>,
+  ): Promise<void> {
+    await super.enrollRemoteRouters(contractsMap);
+
+    this.logger(`Setting enrolled router handle gas overhead...`);
```
Comment (tkporter) [inline typescript/sdk/src/deploy/router/GasRouterDeployer.ts:37]: ditto

File: typescript/sdk/src/deploy/router/GasRouterDeployer.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        remoteDomains.map((domain) => contracts.router.destinationGas(domain)),
+      );
+      const remoteConfigs = remoteDomains
+        .map((domain, i) => ({
+          domain,
+          gas: this.configMap[remoteChains[i]].gas,
+        }))
+        .filter(({ gas }, index) => !currentConfigs[index].eq(gas));
+      this.logger(
+        `Enroll remote (${remoteChains}) handle gas overhead on ${local}`,
```
Comment (tkporter) [inline typescript/sdk/src/deploy/router/GasRouterDeployer.ts:55]: ditto

**Review by tkporter — APPROVED** [review]
nits on possibly missed instances of "overhead"

**Pull request merged**